parcels/particlefile.py

```diff
@@ -31,177 +41,301 @@ def _set_calendar(origin_calendar):
         self.lon.standard_name = "longitude"
         self.lon.units = "degrees_east"
         self.lon.axis = "X"
 
-        self.z = self.dataset.createVariable("z", lonlatdepth_precision, coords, fill_value=np.nan, chunksizes=self.chunksizes)
+        self.z = self.dataset.createVariable("z", lonlatdepth_precision, coords, fill_value=np.nan, chunksizes=data_shape)
         self.z.long_name = ""
         self.z.standard_name = "depth"
         self.z.units = "m"
         self.z.positive = "down"
 
-        self.user_vars = []
-        self.user_vars_once = []
-        """
-        :user_vars: list of additional user defined particle variables to write for all particles and all times
-        :user_vars_once: list of additional user defined particle variables to write for all particles only once at initial time.
-        """
+        for vname in self.var_names:
+            if vname not in ['time', 'lat', 'lon', 'depth', 'id']:
+                setattr(self, vname, self.dataset.createVariable(vname, "f4", coords, fill_value=np.nan, chunksizes=data_shape))
+                getattr(self, vname).long_name = ""
+                getattr(self, vname).standard_name = vname
+                getattr(self, vname).units = "unknown"
+
+        for vname in self.var_names_once:
+            setattr(self, vname, self.dataset.createVariable(vname, "f4", "traj", fill_value=np.nan, chunksizes=[data_shape[0]]))
+            getattr(self, vname).long_name = ""
+            getattr(self, vname).standard_name = vname
+            getattr(self, vname).units = "unknown"
 
-        for v in self.particleset.ptype.variables:
-            if v.name in ['time', 'lat', 'lon', 'depth', 'z', 'id']:
-                continue
-            if v.to_write:
-                if v.to_write is True:
-                    setattr(self, v.name, self.dataset.createVariable(v.name, "f4", coords, fill_value=np.nan, chunksizes=self.chunksizes))
-                    self.user_vars += [v.name]
-                elif v.to_write == 'once':
-                    setattr(self, v.name, self.dataset.createVariable(v.name, "f4", "traj", fill_value=np.nan, chunksizes=[self.chunksizes[0]]))
-                    self.user_vars_once += [v.name]
-                getattr(self, v.name).long_name = ""
-                getattr(self, v.name).standard_name = v.name
-                getattr(self, v.name).units = "unknown"
-
-        self.idx = np.empty(shape=0)
+        for name, message in self.metadata.items():
+            setattr(self.dataset, name, message)
 
     def __del__(self):
-        if self.dataset:
-            self.dataset.close()
+        if self.to_export:
+            self.close()
 
-    def sync(self):
-        """Write all buffered data to disk"""
-        self.dataset.sync()
+    def close(self):
+        """Close the ParticleFile object by exporting and then deleting
+        the temporary npy files"""
+        self.export()
+        self.delete_tempwritedir()
```

Comment:
the temp write dir can sometimes be useful (especially if something goes wrong in a big run)
Should that be an option to delete the temp write dir?
(not sure though, especially since if something goes wrong, the code should break so the dir is not deleted)

Reply:
Yes, indeed it is quite useful that the output-dir exists if the run is somehow broken off. We delete it only after a successful `export()`. 

In the future, we might even want a function to restart a run by creating a particleSet from the last numpy file in a temporary output directory.